sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala

```diff
@@ -33,77 +33,136 @@ import org.apache.spark.sql.rapids.RapidsDiskBlockManager
  */
 class RapidsBufferCatalog extends Logging {
   /** Map of buffer IDs to buffers */
-  private[this] val bufferMap = new ConcurrentHashMap[RapidsBufferId, RapidsBuffer]
+  private[this] val bufferMap = new ConcurrentHashMap[RapidsBufferId, Seq[Option[RapidsBuffer]]]
 
   /**
-   * Lookup the buffer that corresponds to the specified buffer ID and acquire it.
+   * Lookup the buffer that corresponds to the specified buffer ID at the highest storage tier,
+   * and acquire it.
    * NOTE: It is the responsibility of the caller to close the buffer.
    * @param id buffer identifier
    * @return buffer that has been acquired
    */
   def acquireBuffer(id: RapidsBufferId): RapidsBuffer = {
     (0 until RapidsBufferCatalog.MAX_BUFFER_LOOKUP_ATTEMPTS).foreach { _ =>
-      val buffer = bufferMap.get(id)
-      if (buffer == null) {
-        throw new NoSuchElementException(s"Cannot locate buffer associated with ID: $id")
-      }
-      if (buffer.addReference()) {
-        return buffer
+      val buffers = bufferMap.get(id)
+      if (isEmpty(buffers)) {
+        throw new NoSuchElementException(s"Cannot locate buffers associated with ID: $id")
       }
+      buffers.foreach(_.foreach(buffer =>
+        if (buffer.addReference()) {
+          return buffer
+        }
```

Comment:
If we fail to acquire the most appropriate buffer then the buffer was in the process of being freed, meaning the mapping is in the process of being updated.  We need to fully loop back around and fetch a new sequence of buffers and try to acquire the first one as it's likely the sequence has changed.  Only the first one is important for this method.